File: typescript/Helpers/Debug.ts
const flattenStructure = <T>(values: Iterable<T>, separator: string): string => {
  return Array.from(values).join(separator);
};

const isStringArray = (values: unknown[]): values is string[] => values.every((item) => typeof item === "string");

function writeLine(values: Iterable<unknown>): void {
  const entries = Array.from(values);

  if (entries.length === 0) {
    console.log("Empty");
    return;
  }

  if (isStringArray(entries)) {
    console.log(flattenStructure(entries, "\n"));
    return;
  }

  console.log(flattenStructure(entries, ", "));
}

function writeNested<T>(values: Iterable<Iterable<T>>): void {
  const nested = Array.from(values).map((entry) => flattenStructure(entry, ", "));
  writeLine(nested);
}

export const Debug = {
  writeLine,
  writeNested,
};

File: python/helpers/debug.py
from __future__ import annotations

from collections.abc import Iterable
from typing import Any


def _flatten(values: Iterable[Any], separator: str) -> str:
    return separator.join(str(value) for value in values)


def write_line(values: Iterable[Any]) -> None:
    collected = list(values)
    if not collected:
        print("Empty")
        return

    if all(isinstance(item, str) for item in collected):
        print(_flatten(collected, "\n"))
        return

    print(_flatten(collected, ", "))


def write_nested(values: Iterable[Iterable[Any]]) -> None:
    rows = [_flatten(row, ", ") for row in values]
    write_line(rows)


__all__ = ["write_line", "write_nested"]

